engine/crates/engine-v2/schema/src/input_value/de.rs

```diff
@@ -0,0 +1,88 @@
+        IntoDeserializer, Visitor,
+    },
+    forward_to_deserialize_any,
+};
+
+use crate::{InputValue, InputValueSerdeError, InputValuesContext};
+
+pub(super) struct InputValueDeserializer<'de, Str, Ctx> {
+    pub ctx: Ctx,
+    pub value: &'de InputValue<Str>,
+}
+
+impl<'de, Str, Ctx> serde::Deserializer<'de> for InputValueDeserializer<'de, Str, Ctx>
+where
+    Ctx: InputValuesContext<'de, Str>,
+    Str: 'de,
+{
+    type Error = InputValueSerdeError;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        match self.value {
+            InputValue::Null => visitor.visit_unit(),
+            InputValue::String(s) | InputValue::UnknownEnumValue(s) => visitor.visit_borrowed_str(self.ctx.get_str(s)),
+            InputValue::EnumValue(id) => visitor.visit_borrowed_str(self.ctx.schema_walker().walk(*id).name()),
+            InputValue::Int(n) => visitor.visit_i32(*n),
+            InputValue::BigInt(n) => visitor.visit_i64(*n),
+            InputValue::U64(n) => visitor.visit_u64(*n),
+            InputValue::Float(n) => visitor.visit_f64(*n),
+            InputValue::Boolean(b) => visitor.visit_bool(*b),
+            &InputValue::List(ids) => {
+                let ctx = self.ctx;
+                let mut deserializer = SeqDeserializer::new(ids.iter().map(move |id| InputValueDeserializer {
+                    value: &ctx.input_values()[id],
+                    ctx,
+                }));
+                let seq = visitor.visit_seq(&mut deserializer)?;
+                deserializer.end()?;
+                Ok(seq)
+            }
+            InputValue::InputObject(ids) => {
+                let ctx = self.ctx;
+                let mut deserializer = MapDeserializer::new(ids.iter().map(move |id| {
+                    let (id, value) = &ctx.input_values()[id];
+                    (
+                        ctx.schema_walker().walk(*id).name(),
+                        InputValueDeserializer { value, ctx },
+                    )
+                }));
+                let map = visitor.visit_map(&mut deserializer)?;
+                deserializer.end()?;
+                Ok(map)
+            }
+            InputValue::Map(ids) => {
+                let ctx = self.ctx;
+                let mut deserializer = MapDeserializer::new(ids.iter().map(move |id| {
+                    let (key, value) = &ctx.input_values()[id];
+                    (self.ctx.get_str(key), InputValueDeserializer { value, ctx })
+                }));
+                let map = visitor.visit_map(&mut deserializer)?;
+                deserializer.end()?;
+                Ok(map)
+            }
+        }
+    }
+
+    forward_to_deserialize_any! {
+        bool i8 i16 i32 i64 i128 u8 u16 u32 u64 u128 f32 f64 char str string
+        bytes byte_buf option unit unit_struct newtype_struct seq tuple
+        tuple_struct map struct enum identifier ignored_any
+    }
```

Comment:
I faintly recall running into problems doing `forward_to_deserialize_any` like this on option in the past, as some Deserialize impl was expecting a call to `visit_some` on the `Some` path.

Might be worth a unit test to check its ok?

Reply:
I've added a test with a struct here: https://github.com/grafbase/grafbase/pull/1344/files/fd1b73a719fb1b3449dda70caa1ca8a73fd99e9e#diff-ccc4120bf5ff96d2a86233a1448eb1debb054c1d8c469870cef9b6c1d2ecbfefR313

Will add an `Option<String>` in a case where there is actually a string